drivers/timer/rcar_cmt_timer.c

```diff
@@ -0,0 +1,137 @@
+#define CMCNT0_OFFSET                   0x014   /* counter 0 */
+#define CMCSR0_OFFSET                   0x010   /* control/status register 0 */
+
+#define CMCLKE                          0xB00   /* CLK enable register */
+#define CLKEN                           BIT(5)
+
+#define CMSTR0_OFFSET                   0x000   /* Timer start register 0 */
+#define START_BIT                       BIT(0)
+
+#define CSR_CLK_DIV_1                   0x00000007
+#define CSR_ENABLE_COUNTER_IN_DEBUG     BIT(3)
+#define CSR_ENABLE_INTERRUPT            BIT(5)
+#define CSR_FREE_RUN                    BIT(8)
+#define CSR_WRITE_FLAG                  BIT(13)
+#define CSR_OVERFLOW_FLAG               BIT(14)
+#define CSR_MATCH_FLAG                  BIT(15)
+
+static uint32_t cycle_count;
+
+static void cmt_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+	uint32_t reg_val;
+
+	/* clear the interrupt */
+	reg_val = sys_read32(TIMER_BASE_ADDR + CMCSR0_OFFSET);
+	reg_val &= ~CSR_MATCH_FLAG;
+	sys_write32(reg_val, TIMER_BASE_ADDR + CMCSR0_OFFSET);
+
+	cycle_count += CYCLES_PER_TICK;
+
+	/* Announce to the kernel */
+	z_clock_announce(1);
+}
+
+int z_clock_driver_init(const struct device *device)
+{
+	const struct device *clk;
+	uint32_t reg_val;
+	int i, ret;
+
+	clk = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(0));
+
+	ret = clock_control_on(clk, (clock_control_subsys_t *) &mod_clk);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Clock Enable */
+	reg_val = sys_read32(TIMER_BASE_ADDR + CMCLKE);
+	reg_val |= CLKEN;
+	sys_write32(reg_val, TIMER_BASE_ADDR + CMCLKE);
+
+	/* Stop the timer */
+	reg_val = sys_read32(TIMER_BASE_ADDR + CMSTR0_OFFSET);
+	reg_val &= ~START_BIT;
+	sys_write32(reg_val, TIMER_BASE_ADDR + CMSTR0_OFFSET);
+
+	/* Set the timer as 32-bit, with RCLK/1 clock */
+	reg_val = CSR_FREE_RUN;
+	reg_val |= CSR_ENABLE_INTERRUPT;
+	reg_val |= CSR_ENABLE_COUNTER_IN_DEBUG;
+	reg_val |= CSR_CLK_DIV_1;
+	sys_write32(reg_val, TIMER_BASE_ADDR + CMCSR0_OFFSET);
+
+	sys_write32(CYCLES_PER_TICK, TIMER_BASE_ADDR + CMCOR0_OFFSET);
+
+	/* Reset the counter, check WRFLG first */
+	while (sys_read32(TIMER_BASE_ADDR + CMCSR0_OFFSET) & CSR_WRITE_FLAG)
+		;
+	sys_write32(0, TIMER_BASE_ADDR + CMCNT0_OFFSET);
+
+	/* Read this register continuously, until same values are read from this
+	 * register.
+	 */
+	for (i = 0; i < 100; i++) {
+		if (!sys_read32(TIMER_BASE_ADDR + CMCNT0_OFFSET)) {
+			break;
+		}
+	}
+
+	__ASSERT(sys_read32(TIMER_BASE_ADDR + CMCNT0_OFFSET) == 0,
+		 "Fail to clear CMCNT0 register");
+
+	/* Connect timer interrupt */
+	IRQ_CONNECT(TIMER_IRQ, 0, cmt_isr, 0, 0);
+	irq_enable(TIMER_IRQ);
+
+	/* Start the timer */
+	reg_val = sys_read32(TIMER_BASE_ADDR + CMSTR0_OFFSET);
+	reg_val |= START_BIT;
+	sys_write32(reg_val, TIMER_BASE_ADDR + CMSTR0_OFFSET);
+
+	return 0;
+}
+
+uint32_t z_clock_elapsed(void)
+{
+	/* Always return 0 for tickful operation */
+	return 0;
+}
+
+uint32_t z_timer_cycle_get_32(void)
+{
+	return cycle_count;
+}
```

Comment:
Pretty sure the `test_prevent_interrupt` test in `tests/kernel/interrupt` will not pass with this implementation.

Reply:
Correct: by calling irq_lock,  k_busy_wait never return since Timer ISR is not serviced, and so cycle_count is not incremented.
Well the way this timer is working is that the counter is reset to 0 after reaching the match value, and an overflow bit is set.
So at this point I can only add a maximum of CYCLES_PER_TICK to the cycle_count (which is already better), but should not be enough to unlock the k_busy_wait(MS_TO_US(1000));

Reply:
One way to address this issue is, if possible, to return the value of the counter register.

Reply:
Does having k_busy_wait working with irq locked is manadatory ?
In this case I guess this hw timer is not suitable since it just count up to one tick.

Reply:
> Does having k_busy_wait working with irq locked is manadatory ?

A de facto requirement, one could say: https://github.com/zephyrproject-rtos/zephyr/issues/23519

All tests under `tests/` must pass. Try running a full sanitycheck with the real hardware to see if there are any more failures.

Reply:
FWIW: k_busy_wait() needs to work in all contexts, including both locking and high priority ISRs that the timer interrupt can't preempt.  It's used everywhere.  But it doesn't absolutely have to be based on the timer counter, that's just the default.  There are platforms like nRF5 that use a separate mechanism (e.g. another counter device, a cycle-counted loop, etc...).  See CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT.

But that said: long term you really want a tickless-capable timer driver.

Reply:
Thanks for your time and all these advises, I will try to think about a solution.
One of my attempt was to use private timer located in the Cortex-r, I finally gave up with this timer
because it is connected to a different GIC controller than other shared peripherals.
Finally it could be suitable for implementing a custom_busy_wait since it doesn't require interrupts.
One other solution could be to use another channel of this cmt timer that could be kept running, while this one can be kept looping on tick value.

@andyross indeed supporting tickless is the long term solution, in best scenario using the generic arm timer,
if I find a solution for managing to GIC controller within Zephyr.

Reply:
@andyross @stephanosio k_busy_wait issue has been fixed by using a second channel of this timer which is not reset when reaching the tick match value. Tickless support will be added later.